### DISCUSSION TITLE
How can one set the maximum number of training windows in the DeepAREstimator?

### DISCUSSION THREAD

**Discussion by vamp-ire-tap** [awslabs/gluonts#1889] [Q&A] [answered]
Hello, can anyone explain how the training windows are generated in the default call to the DeepAREstimator? I see that on the following line:

https://github.com/awslabs/gluon-ts/blob/8132e2462b649455ade4dbc66cc36114df1796bc/src/gluonts/model/deepar/_estimator.py#L271

There is a call to create the ExpectedNumInstanceSampler(num_instances=1.0, min_future=prediction_length) if there is no train_sampler is provided. By looking at the definition of the ExpectedNumInstanceSampler on the following: 

https://github.com/awslabs/gluon-ts/blob/8132e2462b649455ade4dbc66cc36114df1796bc/src/gluonts/transform/sampler.py#L139

I see that num_instances=1.0 tells the sampling of training windows can occur by choosing any valid point in time series history and sampling some points before and after it to create a training window. However, the main problem I am having is to understand what hyperparameter dictates *how many* training windows should be sampled at max. If we look at the DeepAR paper at https://arxiv.org/pdf/1704.04110.pdf, then in table 3 it is stated that they take #500K training examples for the traffic dataset for example. How can I set this 500K hyperparameter by calling the DeepAREstimator or the ExpectedNumInstanceSampler?

**Comment by vamp-ire-tap** [comment]
Could the authors/contributors of the library answer the question posted above?

**Comment by lostella** [comment, marked as answer]
In general, you can control the maximum number of data samples used for the training procedure by configuring the `Trainer` with
* `epochs`
* `num_batches_per_epoch`

as well as the model (estimator) with the `batch_size`. See [here](https://ts.gluon.ai/api/gluonts/gluonts.mx.trainer.html#module-gluonts.mx.trainer) for the `Trainer` options, and [here](https://ts.gluon.ai/api/gluonts/gluonts.model.deepar.html?highlight=deeparestimator#gluonts.model.deepar.DeepAREstimator) for the estimator options (including `batch_size`).

Once the above options are set, and assuming that early stopping does not kick in, the total number of data samples the training procedure will see is

```
batch_size * num_batches_per_epoch * epochs
```

If early stopping criteria are met, the training loop will stop before having seen that many data stamples.

What you point out above (the `ExpectedNumInstanceSampler`) is definitely related to how data is iterated and sampled: the parameter of `ExpectedNumInstanceSampler` controls how many data samples will be taken on average from each time series as the dataset gets iterated. So if the sampler were to be configured with `2.0` instead of `1.0`, one would observe that (on average) twice as many data samples are sampled before the whole dataset is iterated. I would say that this is more of an internal detail, it regulates "how heavily" to train on samples from each time series before moving to the next ones.

**Reply by vamp-ire-tap** [reply]
thank you very much for a very clear answer. If you don't mind answering a slight tangential question, is the early stopping option set by default, or would one need to configure it through a custom callback?

**Reply by lostella** [reply]
Early stopping is currently enabled by default, through on of the default callbacks. This is mentioned in the documentation for `add_default_callbacks` in the `Trainer` class (linked above in the answer).
